corehq/util/tests/test_couch.py

```diff
@@ -253,7 +253,7 @@ def test_a_few_important_ones(self):
             self.assertEqual(model_class, get_document_class_by_name(model_name))
 
     def test_missing(self):
-        test_cases = [None, '', 'FooDocument']
+        test_cases = [None, 'FooDocument']
```

Comment:
I asked Cory to [review this change](https://github.com/dimagi/commcare-hq/commit/94ba36d1b4681a5d6109c3beba552d1a24c4ca4d#commitcomment-13830855). Also there was a discussion about it in #web-team the other day:

> Daniel Miller [2:31 PM] 
> I'm seeing intermittent failures of `corehq/util/tests/test_couch.py:DocumentClassLookupTest.test_missing`. It appears that some earlier test is creating a document class with no name (empty string). How would I find where that's happening?
> 
> Daniel Miller [2:33 PM]
> `get_document_class_by_name('')` is returning a value instead of raising `DocumentClassNotFound`
> 
> Ben Rudolph [2:34 PM] 
> if you run it by itself it always passes?
> 
> Ben Rudolph [2:34 PM]
> looks like that function is memoized, may want to try reseting the cache on teardown/setup https://github.com/dimagi/dimagi-utils/blob/master/dimagi/utils/decorators/memoized.py#L114
> 
> Ethan Soergel [2:43 PM] 
> `memoized` doesn't use a cache, it stores the computed values on the instance
> 
> Ethan Soergel [2:43 PM]
> I don't think it can be reset, but subsequently instantiated objects won't have that property cached
> 
> Daniel Miller [2:43 PM] 
> yeah, `memoized` is not the problem here.
> 
> Daniel Miller [2:44 PM]
> I stepped into it with pdb and the un-memoized version of `get_document_class_by_name` is returning the class instance
> 
> Ben Rudolph [2:44 PM] 
> ah ok thanks
> 
> Ethan Soergel [2:46 PM] 
> oh wait, `get_document_class_by_name` is a function, not a method - that means its cache will live for the life of the process
> 
> Ethan Soergel [2:48 PM]
> I guess that's kinda the difference between memoization and caching anyways - memoization isn't expected to be invalidated.  I think we blur that line a bit though
> 
> Ben Rudolph [2:51 PM] 
> i do think we can invalidate though
> 
> Ben Rudolph [2:51 PM]
> at least our memoize is able to
> 
> Ethan Soergel [2:52 PM] 
> Daniel, have you tried putting a debugger or print statement in `get_document_class_by_name`?  I think given the actual nameless class object, you should be able to see where it's defined (something like `klass.__module__`, I think)
> 
> Daniel Miller [2:52 PM] 
> I'm trying that now
> 
> Daniel Miller [2:53 PM]
> the returned class is `<class 'jsonobject.base.'>`
> Ethan Soergel 
> [2:55 PM]  

```py
  In [1]: my_class = type("", (), {})

  In [2]: my_class
  Out[2]: __main__.
```

> I'm kinda surprised this is actually possible in Python, actually.
> 
> Ben Rudolph [2:56 PM] 
> ha
> 
> Ethan Soergel [3:01 PM] 
> `JsonObjectMeta.__new__` is likely the culprit.  looks like it asserts that the name is not None, but doesn't assert that it's not "".  You could probably stick a `import traceback; traceback.print_stack()` in there to figure out where it happens
> 
> Ethan Soergel [3:01 PM]
> might also just happen in that file
> 
> Ben Rudolph [3:02 PM] 
> wow nice
> 
> Ethan Soergel [3:02 PM] 
> `JsonObjectMeta` extends `type`, so that's almost certainly involved
> 
> Daniel Miller [3:05 PM] 
> where do you see that "it asserts that the name is not None"? (what line in jsonobject/base.py?) (edited)
> 
> Daniel Miller [3:07 PM]
> Well this is interesting...
> couchdbkit/schema/base.py:50

```
  super(SchemaProperties, mcs).__new__(mcs, '', bases, {})
```

> Ethan Soergel [3:19 PM] 
> oh wow, yeah, that looks like the culprit
> 
> Ethan Soergel [3:19 PM]
> I see that on line 142 in jsonobject/base.py
> 
> Ethan Soergel [3:20 PM]
> oh, but I guess that's a different `name` - I though it was pulling from a parameter dict or something
> Daniel Miller 
> [3:21 PM]  
> corehq/util/tests/test_couch.py

```py
  def test_missing(self):
      test_cases = [None, '', 'FooDocument']

  # CHANGE TO

  def test_missing(self):
      test_cases = [None, 'FooDocument']
```

> Do you think it makes sense to do this?